src/grain/ledger.js

```diff
@@ -0,0 +1,111 @@
+// @flow
+
+import {type NodeAddressT} from "../core/graph";
+import {type Grain} from "./grain";
+
+/**
+ * Tracks a transfer of grain between participants.
+ *
+ * A grainholder may transfer grain to themself (which is a no-op).
+ * A transfer that would result in the sender having a negative balance
+ * is illegal.
+ */
+export type GrainTransferredV1 = {|
+  +type: "GRAIN_TRANSFERRED",
+  +version: number,
+  +sender: NodeAddressT,
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+  +memo: string,
+|};
+
+// TODO: Import GrainAllocation once it merges.
+type GrainAllocation = {};
+
+/**
+ * Tracks a distrbution of newly-minted grain to contributors.
+ *
+ * Each Distribution contains zero or more allocations, which
+ * actually contain the mapping of new grain amounts to contributors.
+ */
+export type GrainDistributedV1 = {|
+  +type: "GRAIN_DISTRIBUTED",
+  +version: number,
+  +timestamMs: number,
+  +memo: string,
+  +allocations: $ReadOnlyArray<GrainAllocation>,
+|};
+
+/**
+ * Record that an "alias" identity is equivalent to a "canonical" identity.
+ *
+ * This event permanently records that one user NodeAddress (the alias) is
+ * equivalent to another NodeAddress (the canonical address). this is useful
+ * if, for example, a user has both a GitHub and a Discourse account. Once the
+ * identities are merged, they cannot be separated (as doing so risks creating
+ * negative balances).
```

Comment:
What about the situation where someone merges an identity that doesn't actually belong to them and then the original owner of the identity wants to claim their identity back?

Reply:
Woah, great question! I hadn't thought of that.

I suppose we could make an "identity unmerge" event that is explicit about how the grain balance gets split between the de-coupled identities. Also now that I think about it, I don't think we would actually get negative balance problems from splitting the identity. However, what would be tricky is retroactively splitting up their combined grain earnings, so we can figure out underpayment properly.

To see an example of when this matters, consider the following attack:
- I merge github/decentralion (which earns most of my "underlying" cred) and sourcecred/decentralion, with sourcecred/decentralion as canonical
- I earn a bunch of grain, paid to sourcecred/decentralion
- I unmerge, a naive algorithm doesn't move around my lifetime earnings to show that a lot of sourcecred/decentralion's earnings now retroactively look like they were earned by github/decentralion
- Now the balanced allocation strategy sees that github/decentralion has a ton of cred but no earnings, so I get a big allocation -- thus double paying me for my cred

So, when doing the unmerge, we would need some way to adjust the history for all the past earnings. We could do this by computing the historical cred scores at the time of the unmerge, looking at the ratio between the different identities cred in each period for which we had a distribution, and re-approportioning each distribution.

Which means the unmerge event would need to include both how to split the grain balance and how to split the earnings, basically. Seems complicated. Let's not implement it for now, but this is a great question to be asking, since we're likely to encounter this in the future.

Reply:
Yea I think its messy to deal with un-merging accounts, even for the future. The ideal way to handle it IMO is actually include "proofs of ownership" as a part of the identity definition. e.g. SourceCred can generate an ID and ask me to tweet it to prove that I own a certain Twitter account. It can ask me to create a gist on GitHub with that ID to prove I own a certain GitHub account. That prevents the ability for anyone to merge an identity that they haven't proved ownership of in the first place.

3Box and Keybase both do this to verify your various social accounts: https://docs.3box.io/faq/data-identity

The reason I even thought to ask this is because we are planning to implement this proof of identity system for MetaGame players so we can automate the whole process instead of manually having to edit the JSON file. MetaGame profiles are all going to be built on 3Box since its fully extensible and designed to be an open identity standard, would be awesome to be able to encode those proofs of identity as a part of SC as well.

Fun fact: The same team behind 3box is now building Ceramic!

Reply:
Unmerging I think will be messy indeed. Although fairly easy to support in the ledger.

It's not unlike a divorce situation. Having used a shared account for some time, it's infeasible to tell who should own what. So you'll have to make some imperfect choice, 50:50? Some other balance? :shrug: you'll have to figure that out _somehow_.

But as far as the ledger is concerned, it's not responsible for how to decide this. It just needs to record the result as an event.

For example `ACCOUNTS_SPLIT` (as the opposite of `ACCOUNTS_MERGED`). It would need to store details about which addresses are involved, and how the Grain is divided. Plus some business rules, like the invariant `sum(grain shares) == original grain` and to undo the the redirecting of aliases.

None of that is rocket science per say on the accounting side. It's the problem of how to divide the Grain that makes it difficult. And I think it's unlikely we'll have a one-size-fits-all solution to support doing this automatically. So crashing here and have an instance maintainer input these things is what I imagine the mid-term solution being.

Reply:
We need to split up both the Grain, and the historical Grain distribution. From perspective of the integrity of the system as a whole, we don't really care who gets the grain, provided that `sum(grain shares) == original grain` gets maintained. However, we do care how the historical grain distribution is attributed. Otherwise you could bind your high-cred account to a low-cred account, and then split accounts in such a way that your historical earnings get attributed to the low-cred account, and then you get to reap double distributions when the lifetime payout strategy sees that your high-cred account is underpaid.

The correct way to do this is to know the Cred scores *after* the split has been applied, and use that as a basis for computing the ledger update. But that is kind of tricky, because I'd like to have the Cred score computation pipeline use the Ledger as SoT for identities, but actually you'd need to compute Cred on the new (split) identities to get the data needed to properly split them.